File: src/kube_chainsaw/analyzer.py
"""RBAC analyzer with 15 detection rules (KC-001 through KC-015)."""

from collections import defaultdict
from typing import Dict, List, Optional, Set, Tuple

from .models import AnalyzerError, BindingScope, Finding, LoadedResources, Severity


# Dangerous verbs that indicate privilege escalation risk
DANGEROUS_VERBS: Dict[str, str] = {
    "*": "KC-002",
    "escalate": "KC-003",
    "impersonate": "KC-004",
    "bind": "KC-005",
}

# Dangerous resources that indicate access to sensitive data or control plane
DANGEROUS_RESOURCES: Dict[str, str] = {
    "*": "KC-001",
    "secrets": "KC-006",
    "pods/exec": "KC-007",
    "pods/attach": "KC-007",
    "nodes": "KC-008",
    "persistentvolumes": "KC-009",
    "clusterroles": "KC-010",
    "clusterrolebindings": "KC-010",
}

# Escalation combination: verbs X resources in the same rule
ESCALATION_COMBOS: List[Tuple[Set[str], Set[str], str]] = [
    (
        {"create", "patch", "update"},
        {"roles", "clusterroles", "rolebindings", "clusterrolebindings"},
        "KC-011",
    ),
    (
        {"create"},
        {"pods"},
        "KC-012",
    ),
]


class RBACAnalyzer:
    """Analyzes Kubernetes RBAC manifests for security issues.

    Usage:
        analyzer = RBACAnalyzer()
        analyzer.load(resources)
        findings = analyzer.analyze()
    """

    def __init__(self) -> None:
        self._resources: Optional[LoadedResources] = None
        self._findings: List[Finding] = []

    def load(self, resources: LoadedResources) -> None:
        """Load resources for analysis.

        Args:
            resources: Loaded Kubernetes resources from load_manifests().
        """
        self._resources = resources

    def analyze(self) -> List[Finding]:
        """Run all 15 detection rules and return findings.

        Returns:
            List of Finding objects, one per triggered rule per resource.

        Raises:
            AnalyzerError: If load() has not been called.
        """
        if self._resources is None:
            raise AnalyzerError("load() must be called before analyze()")

        self._findings = []

        # Analyze ClusterRoles for dangerous permissions (KC-001 through KC-012)
        self._analyze_cluster_roles()

        # Analyze Roles for dangerous permissions (same rules, severity capped)
        self._analyze_roles()

        # Check aggregated ClusterRoles (KC-015)
        self._check_aggregated_roles()

        # Analyze privilege chains: SA -> Pod mappings (KC-013, KC-014)
        self._analyze_privilege_chains()

        return self._findings

    # ------------------------------------------------------------------
    # Binding scope resolution
    # ------------------------------------------------------------------

    def _get_binding_scope(self, role_name: str) -> BindingScope:
        """Determine if a ClusterRole is bound cluster-wide or namespace-scoped.

        Checks both ClusterRoleBindings and RoleBindings that reference
        this ClusterRole by name.

        Args:
            role_name: Name of the ClusterRole.

        Returns:
            BindingScope with cluster_wide, namespace_scoped, unbound flags
            and subject type counts.
        """
        assert self._resources is not None

        cluster_bindings: List[str] = []
        for binding in self._resources.cluster_role_bindings:
            role_ref = binding["doc"].get("roleRef", {})
            if role_ref.get("name") == role_name:
                binding_name = binding["doc"].get("metadata", {}).get("name", "unknown")
                cluster_bindings.append(binding_name)

        role_bindings: List[str] = []
        for binding in self._resources.role_bindings:
            role_ref = binding["doc"].get("roleRef", {})
            if role_ref.get("name") == role_name and role_ref.get("kind") == "ClusterRole":
                binding_name = binding["doc"].get("metadata", {}).get("name", "unknown")
                role_bindings.append(binding_name)

        # Count subject types across all bindings referencing this ClusterRole
        subject_types: Dict[str, int] = {"ServiceAccount": 0, "Group": 0, "User": 0}
        for binding in self._resources.cluster_role_bindings + self._resources.role_bindings:
            role_ref = binding["doc"].get("roleRef", {})
            if role_ref.get("name") == role_name and role_ref.get("kind") == "ClusterRole":
                for subject in binding["doc"].get("subjects", []):
                    kind = subject.get("kind", "")
                    if kind in subject_types:
                        subject_types[kind] += 1

        return BindingScope(
            cluster_wide=len(cluster_bindings) > 0,
            namespace_scoped=len(role_bindings) > 0,
            unbound=len(cluster_bindings) == 0 and len(role_bindings) == 0,
            cluster_bindings=cluster_bindings,
            role_bindings=role_bindings,
            subject_types=subject_types,
        )

    # ------------------------------------------------------------------
    # Severity determination
    # ------------------------------------------------------------------

    def _determine_severity(
        self,
        binding_scope: BindingScope,
        has_wildcard: bool,
    ) -> Severity:
        """Determine finding severity based on binding scope and permissions.

        Severity ladder:
        - Unbound ClusterRole -> INFO
        - Namespace-scoped without wildcards -> WARNING
        - Namespace-scoped with wildcards -> HIGH
        - Cluster-wide without wildcards -> HIGH
        - Cluster-wide with wildcards -> CRITICAL

        When a ClusterRole has both ClusterRoleBindings AND RoleBindings,
        the cluster-wide scope takes precedence (highest scope wins).

        Args:
            binding_scope: Binding scope for the ClusterRole.
            has_wildcard: Whether the triggering rule contains wildcard
                          verbs or resources.

        Returns:
            Severity enum value.
        """
        if binding_scope.unbound:
            return Severity.INFO

        if binding_scope.cluster_wide:
            if has_wildcard:
                return Severity.CRITICAL
            return Severity.HIGH

        # namespace_scoped only
        if has_wildcard:
            return Severity.HIGH
        return Severity.WARNING

    # ------------------------------------------------------------------
    # ClusterRole analysis
    # ------------------------------------------------------------------

    def _analyze_cluster_roles(self) -> None:
        """Analyze ClusterRoles for dangerous verbs, resources, and escalation combos."""
        assert self._resources is not None

        for role_name, role_data in self._resources.cluster_roles.items():
            rules = role_data.get("rules", [])
            file_path = role_data.get("file", "")

            binding_scope = self._get_binding_scope(role_name)

            # Track which rule_ids have been triggered for this role
            # to avoid duplicate findings for the same rule_id on the same resource
            triggered_rule_ids: Set[str] = set()

            # Determine if any rule in this role has wildcards (for severity calc)
            has_any_wildcard = False
            for rule in rules:
                resources = [r.lower() for r in rule.get("resources", [])]
                verbs = [v.lower() for v in rule.get("verbs", [])]
                if "*" in resources or "*" in verbs:
                    has_any_wildcard = True
                    break

            for rule in rules:
                resources = [r.lower() for r in rule.get("resources", [])]
                verbs = [v.lower() for v in rule.get("verbs", [])]

                # Check dangerous verbs
                for verb in verbs:
                    rule_id = DANGEROUS_VERBS.get(verb)
                    if rule_id and rule_id not in triggered_rule_ids:
                        triggered_rule_ids.add(rule_id)
                        # Wildcard check: is this specific finding about a wildcard?
                        is_wildcard = verb == "*"
                        severity = self._determine_severity(binding_scope, has_any_wildcard or is_wildcard)
                        self._findings.append(Finding(
                            rule_id=rule_id,
                            severity=severity,
                            title=f"ClusterRole {role_name} has dangerous verb: {verb}",
                            file=file_path,
                            description=f"Verb '{verb}' in ClusterRole {role_name}",
                            remediation="Remove dangerous verb or scope to minimum required permissions",
                            resource_kind="ClusterRole",
                            resource_name=role_name,
                        ))

                # Check dangerous resources
                for resource in resources:
                    rule_id = DANGEROUS_RESOURCES.get(resource)
                    if rule_id and rule_id not in triggered_rule_ids:
                        triggered_rule_ids.add(rule_id)
                        is_wildcard = resource == "*"
                        severity = self._determine_severity(binding_scope, has_any_wildcard or is_wildcard)
                        self._findings.append(Finding(
                            rule_id=rule_id,
                            severity=severity,
                            title=f"ClusterRole {role_name} has dangerous resource: {resource}",
                            file=file_path,
                            description=f"Resource '{resource}' in ClusterRole {role_name}",
                            remediation="Remove dangerous resource access or scope to minimum required permissions",
                            resource_kind="ClusterRole",
                            resource_name=role_name,
                        ))

                # Check escalation combos (verb + resource in same rule)
                for esc_verbs, esc_resources, rule_id in ESCALATION_COMBOS:
                    if rule_id in triggered_rule_ids:
                        continue
                    # Both sets must have intersection in the SAME rule
                    if set(verbs) & esc_verbs and set(resources) & esc_resources:
                        triggered_rule_ids.add(rule_id)
                        severity = self._determine_severity(binding_scope, has_any_wildcard)
                        self._findings.append(Finding(
                            rule_id=rule_id,
                            severity=severity,
                            title=f"ClusterRole {role_name} has escalation combo",
                            file=file_path,
                            description=(
                                f"Escalation risk: verbs {sorted(set(verbs) & esc_verbs)} "
                                f"on resources {sorted(set(resources) & esc_resources)} "
                                f"in ClusterRole {role_name}"
                            ),
                            remediation="Remove escalation-enabling verb/resource combination",
                            resource_kind="ClusterRole",
                            resource_name=role_name,
                        ))

    # ------------------------------------------------------------------
    # Role analysis (namespace-scoped, severity capped at WARNING)
    # ------------------------------------------------------------------

    def _analyze_roles(self) -> None:
        """Analyze namespace-scoped Roles for dangerous permissions.

        Same detection rules as ClusterRoles, but severity is capped at WARNING.
        """
        assert self._resources is not None

        for role_key, role_data in self._resources.roles.items():
            rules = role_data.get("rules", [])
            file_path = role_data.get("file", "")

            # role_key format: "namespace/name"
            parts = role_key.split("/", 1)
            namespace = parts[0] if len(parts) > 1 else "default"
            role_name = parts[1] if len(parts) > 1 else parts[0]

            # Check if this Role is bound (has a RoleBinding referencing it)
            is_bound = False
            for binding in self._resources.role_bindings:
                role_ref = binding["doc"].get("roleRef", {})
                if role_ref.get("name") == role_name and role_ref.get("kind") == "Role":
                    bind_ns = binding["doc"].get("metadata", {}).get("namespace", "default")
                    if bind_ns == namespace:
                        is_bound = True
                        break

            triggered_rule_ids: Set[str] = set()

            for rule in rules:
                resources = [r.lower() for r in rule.get("resources", [])]
                verbs = [v.lower() for v in rule.get("verbs", [])]

                # Check dangerous verbs
                for verb in verbs:
                    rule_id = DANGEROUS_VERBS.get(verb)
                    if rule_id and rule_id not in triggered_rule_ids:
                        triggered_rule_ids.add(rule_id)
                        severity = Severity.WARNING if is_bound else Severity.INFO
                        self._findings.append(Finding(
                            rule_id=rule_id,
                            severity=severity,
                            title=f"Role {role_name} has dangerous verb: {verb}",
                            file=file_path,
                            description=f"Verb '{verb}' in Role {namespace}/{role_name}",
                            remediation="Remove dangerous verb or scope to minimum required permissions",
                            resource_kind="Role",
                            resource_name=role_name,
                            resource_namespace=namespace,
                        ))

                # Check dangerous resources
                for resource in resources:
                    rule_id = DANGEROUS_RESOURCES.get(resource)
                    if rule_id and rule_id not in triggered_rule_ids:
                        triggered_rule_ids.add(rule_id)
                        severity = Severity.WARNING if is_bound else Severity.INFO
                        self._findings.append(Finding(
                            rule_id=rule_id,
                            severity=severity,
                            title=f"Role {role_name} has dangerous resource: {resource}",
                            file=file_path,
                            description=f"Resource '{resource}' in Role {namespace}/{role_name}",
                            remediation="Remove dangerous resource access or scope to minimum required permissions",
                            resource_kind="Role",
                            resource_name=role_name,
                            resource_namespace=namespace,
                        ))

                # Check escalation combos
                for esc_verbs, esc_resources, rule_id in ESCALATION_COMBOS:
                    if rule_id in triggered_rule_ids:
                        continue
                    if set(verbs) & esc_verbs and set(resources) & esc_resources:
                        triggered_rule_ids.add(rule_id)
                        severity = Severity.WARNING if is_bound else Severity.INFO
                        self._findings.append(Finding(
                            rule_id=rule_id,
                            severity=severity,
                            title=f"Role {role_name} has escalation combo",
                            file=file_path,
                            description=(
                                f"Escalation risk: verbs {sorted(set(verbs) & esc_verbs)} "
                                f"on resources {sorted(set(resources) & esc_resources)} "
                                f"in Role {namespace}/{role_name}"
                            ),
                            remediation="Remove escalation-enabling verb/resource combination",
                            resource_kind="Role",
                            resource_name=role_name,
                            resource_namespace=namespace,
                        ))

    # ------------------------------------------------------------------
    # Aggregated ClusterRoles (KC-015)
    # ------------------------------------------------------------------

    def _check_aggregated_roles(self) -> None:
        """Detect aggregated ClusterRoles (KC-015), always INFO severity."""
        assert self._resources is not None

        for role_name, role_data in self._resources.cluster_roles.items():
            doc = role_data.get("doc", {})
            if "aggregationRule" in doc:
                selectors = doc["aggregationRule"].get("clusterRoleSelectors", [])
                self._findings.append(Finding(
                    rule_id="KC-015",
                    severity=Severity.INFO,
                    title=f"Aggregated ClusterRole detected: {role_name}",
                    file=role_data.get("file", ""),
                    description=f"Aggregates permissions from roles matching {selectors}",
                    remediation="Review aggregation selectors to ensure no unintended permissions are granted",
                    resource_kind="ClusterRole",
                    resource_name=role_name,
                ))

    # ------------------------------------------------------------------
    # Privilege chain analysis (KC-013, KC-014)
    # ------------------------------------------------------------------

    def _analyze_privilege_chains(self) -> None:
        """Build ClusterRole -> Binding -> SA -> Pod chains.

        KC-013: Pod with SA bound to cluster-admin.
        KC-014: RoleBinding (not ClusterRoleBinding) referencing a ClusterRole,
                where a Pod uses that SA.
        """
        assert self._resources is not None

        # Build SA -> permission mappings
        sa_permissions: Dict[str, List[dict]] = defaultdict(list)

        # From ClusterRoleBindings
        for binding in self._resources.cluster_role_bindings:
            doc = binding["doc"]
            role_ref = doc.get("roleRef", {})
            role_name = role_ref.get("name", "")
            for subject in doc.get("subjects", []):
                if subject.get("kind") == "ServiceAccount":
                    sa_ns = subject.get("namespace", "default")
                    sa_name = subject.get("name", "")
                    sa_key = f"{sa_ns}/{sa_name}"
                    sa_permissions[sa_key].append({
                        "type": "ClusterRoleBinding",
                        "role_kind": role_ref.get("kind", "ClusterRole"),
                        "role": role_name,
                        "binding": doc.get("metadata", {}).get("name", "unknown"),
                        "file": binding["file"],
                    })

        # From RoleBindings
        for binding in self._resources.role_bindings:
            doc = binding["doc"]
            role_ref = doc.get("roleRef", {})
            role_kind = role_ref.get("kind", "")
            role_name = role_ref.get("name", "")
            binding_ns = doc.get("metadata", {}).get("namespace", "default")
            for subject in doc.get("subjects", []):
                if subject.get("kind") == "ServiceAccount":
                    sa_ns = subject.get("namespace", binding_ns)
                    sa_name = subject.get("name", "")
                    sa_key = f"{sa_ns}/{sa_name}"
                    sa_permissions[sa_key].append({
                        "type": "RoleBinding",
                        "role_kind": role_kind,
                        "role": role_name,
                        "binding": doc.get("metadata", {}).get("name", "unknown"),
                        "file": binding["file"],
                    })

        # Track already-reported bindings for KC-014 dedup
        reported_bindings: Set[str] = set()

        # Map Pods to their SA permissions
        for pod_key, pod_info in self._resources.pods.items():
            namespace = pod_key.split("/")[0]
            pod_name = pod_key.split("/")[1] if "/" in pod_key else pod_key
            sa_name = pod_info.get("serviceAccount", "default")
            sa_key = f"{namespace}/{sa_name}"

            if sa_key not in sa_permissions:
                continue

            for perm in sa_permissions[sa_key]:
                role_name = perm["role"]

                # KC-013: Pod with cluster-admin
                if role_name == "cluster-admin":
                    self._findings.append(Finding(
                        rule_id="KC-013",
                        severity=Severity.CRITICAL,
                        title=f"Pod {pod_key} has cluster-admin access",
                        file=pod_info.get("file", ""),
                        description=(
                            f"Pod uses ServiceAccount {sa_key} bound to cluster-admin"
                        ),
                        remediation="Create a custom Role with minimal required permissions",
                        resource_kind="Pod",
                        resource_name=pod_name,
                        resource_namespace=namespace,
                    ))

                # KC-014: RoleBinding referencing a ClusterRole, where a Pod uses that SA
                if (
                    perm["type"] == "RoleBinding"
                    and perm["role_kind"] == "ClusterRole"
                    and perm["binding"] not in reported_bindings
                ):
                    reported_bindings.add(perm["binding"])
                    self._findings.append(Finding(
                        rule_id="KC-014",
                        severity=Severity.WARNING,
                        title=f"RoleBinding {perm['binding']} grants cluster-wide permissions",
                        file=perm["file"],
                        description=(
                            f"RoleBinding references ClusterRole {role_name}, "
                            f"granting cluster-scoped permissions in namespace scope"
                        ),
                        remediation="Use a namespace-scoped Role instead of ClusterRole",
                        resource_kind="Pod",
                        resource_name=pod_name,
                        resource_namespace=namespace,
                    ))

File: pkg/analyzer/analyzer.go
package analyzer

import (
	"fmt"
	"strings"

	"github.com/ugiordan/kube-chainsaw/pkg/models"
)

// Analyze examines loaded RBAC resources and returns security findings.
func Analyze(resources *models.LoadedResources) []models.Finding {
	if resources == nil || resources.IsEmpty() {
		return nil
	}

	var findings []models.Finding

	// Phase 1: Analyze ClusterRoles
	for name, cr := range resources.ClusterRoles {
		scope := resolveClusterRoleScope(name, resources)

		// Check for aggregated ClusterRole (KC-015)
		if isAggregated(cr.Doc) {
			f := newFinding(RuleAggregatedClusterRole, models.SeverityInfo, cr.File, "ClusterRole", name, "")
			findings = appendIfNew(findings, f)
		}

		// Check rules for dangerous patterns
		findings = append(findings, checkRules(cr.Rules, name, "ClusterRole", "", cr.File, &scope, false)...)
	}

	// Phase 2: Analyze Roles (namespace-scoped, severity capped at WARNING)
	for key, role := range resources.Roles {
		scope := resolveRoleScope(key, resources)
		findings = append(findings, checkRules(role.Rules, extractNameFromKey(key), "Role", role.Namespace, role.File, &scope, true)...)
	}

	// Phase 3: Privilege chain analysis
	findings = append(findings, analyzePrivilegeChains(resources)...)

	return findings
}

// checkRules inspects the rules of a role (ClusterRole or Role) for dangerous patterns.
func checkRules(rules []map[string]interface{}, roleName, roleKind, namespace, file string, scope *models.BindingScope, isNamespaced bool) []models.Finding {
	var findings []models.Finding
	seen := make(map[string]bool) // deduplicate: rule_id per resource

	for _, rule := range rules {
		verbs := toStringSlice(rule["verbs"])
		resources := toStringSlice(rule["resources"])

		hasWildcardVerb := contains(verbs, "*")
		hasWildcardResource := contains(resources, "*")
		hasWildcards := hasWildcardVerb || hasWildcardResource

		// Check dangerous verbs
		for _, verb := range verbs {
			if ruleID, ok := dangerousVerbs[verb]; ok {
				dedup := ruleID + "|" + roleName
				if !seen[dedup] {
					seen[dedup] = true
					sev := computeSeverity(scope, hasWildcards)
					if isNamespaced {
						sev = capSeverity(sev, models.SeverityWarning)
					}
					f := newFinding(ruleID, sev, file, roleKind, roleName, namespace)
					f.Description = fmt.Sprintf("Role %q has dangerous verb %q", roleName, verb)
					findings = append(findings, f)
				}
			}
		}

		// Check dangerous resources
		for _, res := range resources {
			if ruleID, ok := dangerousResources[res]; ok {
				dedup := ruleID + "|" + roleName
				if !seen[dedup] {
					seen[dedup] = true
					sev := computeSeverity(scope, hasWildcards)
					if isNamespaced {
						sev = capSeverity(sev, models.SeverityWarning)
					}
					f := newFinding(ruleID, sev, file, roleKind, roleName, namespace)
					f.Description = fmt.Sprintf("Role %q grants access to dangerous resource %q", roleName, res)
					findings = append(findings, f)
				}
			}
		}

		// Check escalation combos: KC-011 (create/patch/update on roles/bindings)
		if hasEscalationBindingCombo(verbs, resources) {
			dedup := RuleEscalationBindings + "|" + roleName
			if !seen[dedup] {
				seen[dedup] = true
				sev := computeSeverity(scope, hasWildcards)
				if isNamespaced {
					sev = capSeverity(sev, models.SeverityWarning)
				}
				f := newFinding(RuleEscalationBindings, sev, file, roleKind, roleName, namespace)
				f.Description = fmt.Sprintf("Role %q can create/modify roles or bindings (privilege escalation risk)", roleName)
				findings = append(findings, f)
			}
		}

		// Check escalation combo: KC-012 (create on pods/workloads)
		if hasEscalationPodCombo(verbs, resources) {
			dedup := RuleEscalationPodCreation + "|" + roleName
			if !seen[dedup] {
				seen[dedup] = true
				sev := computeSeverity(scope, hasWildcards)
				if isNamespaced {
					sev = capSeverity(sev, models.SeverityWarning)
				}
				f := newFinding(RuleEscalationPodCreation, sev, file, roleKind, roleName, namespace)
				f.Description = fmt.Sprintf("Role %q can create pods/workloads (privilege escalation risk)", roleName)
				findings = append(findings, f)
			}
		}
	}

	return findings
}

// hasEscalationBindingCombo returns true if verbs include create/patch/update AND
// resources include roles, clusterroles, rolebindings, or clusterrolebindings.
func hasEscalationBindingCombo(verbs, resources []string) bool {
	hasMutationVerb := false
	for _, v := range verbs {
		if escalationMutationVerbs[v] || v == "*" {
			hasMutationVerb = true
			break
		}
	}
	if !hasMutationVerb {
		return false
	}

	for _, r := range resources {
		if escalationBindingResources[r] || r == "*" {
			return true
		}
	}
	return false
}

// hasEscalationPodCombo returns true if verbs include "create" (or "*") AND
// resources include pods or workload controllers.
func hasEscalationPodCombo(verbs, resources []string) bool {
	hasCreate := false
	for _, v := range verbs {
		if v == "create" || v == "*" {
			hasCreate = true
			break
		}
	}
	if !hasCreate {
		return false
	}

	for _, r := range resources {
		if escalationPodResources[r] || r == "*" {
			return true
		}
	}
	return false
}

// analyzePrivilegeChains walks Pod -> SA -> Binding -> Role chains.
func analyzePrivilegeChains(resources *models.LoadedResources) []models.Finding {
	var findings []models.Finding

	for _, pod := range resources.Pods {
		saKey := pod.Namespace + "/" + pod.ServiceAccountName

		// Check ClusterRoleBindings referencing this SA
		for _, crb := range resources.ClusterRoleBindings {
			if !bindingReferencesSubject(crb, pod.ServiceAccountName, pod.Namespace) {
				continue
			}

			roleRefName := getRoleRefName(crb.RoleRef)
			roleRefKind := getRoleRefKind(crb.RoleRef)

			if roleRefKind == "ClusterRole" && roleRefName == "cluster-admin" {
				// KC-013: Pod running with cluster-admin
				f := newFinding(RuleClusterAdminPod, models.SeverityCritical, pod.File, "Pod", pod.Name, pod.Namespace)
				f.Description = fmt.Sprintf(
					"Pod %q uses ServiceAccount %q which is bound to cluster-admin via ClusterRoleBinding %q",
					pod.Name, saKey, crb.Name,
				)
				findings = appendIfNew(findings, f)
			}
		}

		// Check RoleBindings referencing a ClusterRole (KC-014)
		for _, rb := range resources.RoleBindings {
			if !bindingReferencesSubject(rb, pod.ServiceAccountName, pod.Namespace) {
				continue
			}

			roleRefKind := getRoleRefKind(rb.RoleRef)
			if roleRefKind == "ClusterRole" {
				roleRefName := getRoleRefName(rb.RoleRef)
				f := newFinding(RuleRoleBindingClusterRef, models.SeverityWarning, rb.File, "RoleBinding", rb.Name, rb.Namespace)
				f.Description = fmt.Sprintf(
					"RoleBinding %q in namespace %q references ClusterRole %q (used by pod %q via SA %q)",
					rb.Name, rb.Namespace, roleRefName, pod.Name, saKey,
				)
				findings = appendIfNew(findings, f)
			}
		}
	}

	// Also check RoleBindings referencing ClusterRoles even without pods
	for _, rb := range resources.RoleBindings {
		roleRefKind := getRoleRefKind(rb.RoleRef)
		if roleRefKind == "ClusterRole" {
			roleRefName := getRoleRefName(rb.RoleRef)
			f := newFinding(RuleRoleBindingClusterRef, models.SeverityWarning, rb.File, "RoleBinding", rb.Name, rb.Namespace)
			f.Description = fmt.Sprintf(
				"RoleBinding %q in namespace %q references ClusterRole %q",
				rb.Name, rb.Namespace, roleRefName,
			)
			findings = appendIfNew(findings, f)
		}
	}

	return findings
}

// resolveClusterRoleScope determines how a ClusterRole is bound.
func resolveClusterRoleScope(name string, resources *models.LoadedResources) models.BindingScope {
	scope := models.BindingScope{
		SubjectTypes: make(map[string]int),
	}

	// Check ClusterRoleBindings
	for _, crb := range resources.ClusterRoleBindings {
		if getRoleRefName(crb.RoleRef) == name && getRoleRefKind(crb.RoleRef) == "ClusterRole" {
			scope.ClusterWide = true
			scope.ClusterBindings = append(scope.ClusterBindings, crb.Name)
			countSubjects(crb.Subjects, scope.SubjectTypes)
		}
	}

	// Check RoleBindings referencing this ClusterRole
	for _, rb := range resources.RoleBindings {
		if getRoleRefName(rb.RoleRef) == name && getRoleRefKind(rb.RoleRef) == "ClusterRole" {
			scope.NamespaceScoped = true
			scope.RoleBindings = append(scope.RoleBindings, rb.Name)
			countSubjects(rb.Subjects, scope.SubjectTypes)
		}
	}

	if !scope.ClusterWide && !scope.NamespaceScoped {
		scope.Unbound = true
	}

	return scope
}

// resolveRoleScope determines how a Role is bound.
func resolveRoleScope(key string, resources *models.LoadedResources) models.BindingScope {
	scope := models.BindingScope{
		SubjectTypes: make(map[string]int),
	}

	role, ok := resources.Roles[key]
	if !ok {
		scope.Unbound = true
		return scope
	}

	name := extractNameFromKey(key)

	for _, rb := range resources.RoleBindings {
		if getRoleRefName(rb.RoleRef) == name && getRoleRefKind(rb.RoleRef) == "Role" && rb.Namespace == role.Namespace {
			scope.NamespaceScoped = true
			scope.RoleBindings = append(scope.RoleBindings, rb.Name)
			countSubjects(rb.Subjects, scope.SubjectTypes)
		}
	}

	if !scope.NamespaceScoped {
		scope.Unbound = true
	}

	return scope
}

// Helper functions

func newFinding(ruleID string, severity models.Severity, file, resourceKind, resourceName, namespace string) models.Finding {
	f := models.Finding{
		RuleID:            ruleID,
		Severity:          severity,
		Title:             ruleDescriptions[ruleID],
		File:              file,
		Remediation:       ruleRemediations[ruleID],
		ResourceKind:      resourceKind,
		ResourceName:      resourceName,
		ResourceNamespace: namespace,
	}
	f.ComputeFingerprint()
	return f
}

func appendIfNew(findings []models.Finding, f models.Finding) []models.Finding {
	for _, existing := range findings {
		if existing.Fingerprint == f.Fingerprint {
			return findings
		}
	}
	return append(findings, f)
}

func capSeverity(s, max models.Severity) models.Severity {
	if s > max {
		return max
	}
	return s
}

func toStringSlice(v interface{}) []string {
	if v == nil {
		return nil
	}
	switch val := v.(type) {
	case []interface{}:
		result := make([]string, 0, len(val))
		for _, item := range val {
			if s, ok := item.(string); ok {
				result = append(result, s)
			}
		}
		return result
	case []string:
		return val
	default:
		return nil
	}
}

func contains(slice []string, val string) bool {
	for _, s := range slice {
		if s == val {
			return true
		}
	}
	return false
}

func isAggregated(doc map[string]interface{}) bool {
	_, ok := doc["aggregationRule"]
	return ok
}

func getRoleRefName(roleRef map[string]interface{}) string {
	if roleRef == nil {
		return ""
	}
	name, _ := roleRef["name"].(string)
	return name
}

func getRoleRefKind(roleRef map[string]interface{}) string {
	if roleRef == nil {
		return ""
	}
	kind, _ := roleRef["kind"].(string)
	return kind
}

func bindingReferencesSubject(binding *models.BindingData, saName, saNamespace string) bool {
	for _, subj := range binding.Subjects {
		kind, _ := subj["kind"].(string)
		name, _ := subj["name"].(string)
		ns, _ := subj["namespace"].(string)

		if kind == "ServiceAccount" && name == saName && ns == saNamespace {
			return true
		}
	}
	return false
}

func countSubjects(subjects []map[string]interface{}, counts map[string]int) {
	for _, subj := range subjects {
		kind, _ := subj["kind"].(string)
		if kind != "" {
			counts[kind]++
		}
	}
}

func extractNameFromKey(key string) string {
	parts := strings.SplitN(key, "/", 2)
	if len(parts) == 2 {
		return parts[1]
	}
	return key
}
